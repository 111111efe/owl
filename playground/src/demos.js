export const HELLO_WORLD = `class HelloWorld extends owl.core.Component {
  inlineTemplate = \`<div>Hello <t t-esc="props.name"/></div>\`;
}

const env = {
    qweb: new owl.core.QWeb()
};

const hello = new HelloWorld(env, { name: "World" });
hello.mount(document.body);
`;
