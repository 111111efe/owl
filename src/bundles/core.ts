import QWeb from "../core/qweb_vdom";
import * as utils from "../core/utils";
import Widget from "../core/widget";


(<any>window).core = {
    QWeb,
    utils,
    Widget
}