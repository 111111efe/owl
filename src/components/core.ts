import { qweb } from "../qweb/qweb";
import { QWebTemplate } from "../qweb/types";
import { VRootNode } from "../vdom/types";
import { buildTree, patch } from "../vdom/vdom";
import { Component } from "./component";
import { Fiber } from "./fiber";
import { scheduler } from "./scheduler";

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

export interface FComponent<T = any, Env = any> {
  template: string;
  components?: { [key: string]: FComponent<any> | CComponent };
  name?: string;
  setup?: (props: any, env: Env) => T | Promise<T> | void;
}

export type CComponent = typeof Component;

export type CInstance<Props = any, Env = any> = Component<Props, Env>;

export type FInstance<T = {}, Env = any> = T & {
  __owl__: OwlElement<T>;
  props: any;
  env: Env;
};

export interface OwlElement<T = any> {
  fiber: Fiber | null;
  name: string;
  components: { [key: string]: FComponent<T> | CComponent };
  isMounted: boolean;
  instance: FInstance<T> | CInstance | null;
  vnode: VRootNode | null;
  qweb: QWebTemplate;
  isReady: Promise<void> | null;
  children: { [key: string]: OwlElement };
}

// -----------------------------------------------------------------------------
// core object
// -----------------------------------------------------------------------------
export const core = {
  current: null as null | OwlElement,
};

// -----------------------------------------------------------------------------
// Mount
// -----------------------------------------------------------------------------
export type MountTarget = HTMLElement | DocumentFragment;

export interface MountConfig<Env = any> {
  props?: Object;
  env?: Env;
}

export function mount<C extends CComponent>(
  Comp: C,
  target: MountTarget,
  config?: MountConfig
): Promise<InstanceType<C>>;
export function mount(
  comp: CInstance,
  target: MountTarget,
  config?: MountConfig
): Promise<CInstance>;
export function mount<T, Env>(
  Comp: FComponent<T>,
  target: MountTarget,
  config?: MountConfig<Env>
): Promise<FInstance<T extends {} ? T : {}, Env>>;
export function mount<T, Env>(
  comp: FInstance,
  target: MountTarget,
  config?: MountConfig<Env>
): Promise<FInstance<T extends {} ? T : {}, Env>>;
export async function mount(
  info: any,
  target: MountTarget,
  config: MountConfig = {}
): Promise<CInstance | FInstance> {
  if (!(target instanceof HTMLElement || target instanceof DocumentFragment)) {
    const name =
      info instanceof Component ? info.constructor.name : (info as any).name || "No Name";
    let message = `Component '${name}' cannot be mounted: the target is not a valid DOM node.`;
    message += `\nMaybe the DOM is not ready yet? (in that case, you can use owl.utils.whenReady)`;
    throw new Error(message);
  }
  if (info instanceof Component) {
    // this is a CInstance
    const elem = info.__owl__!;
    return scheduler.addFiber(elem.fiber!).then(() => {
      target.appendChild(info.el!);
      return info;
    });
  }
  let element: OwlElement;
  let props = config.props || {};
  let env = config.env || {};
  if (info.prototype instanceof Component || info === Component) {
    // this is a CComponent (Class Component)
    element = createCComponent(info, props, env);
  } else if ("__owl__" in info) {
    // this is a CInstance or a FInstance
    element = info.__owl__;
  } else {
    // this is a FComponent
    element = createFComponent(info, props, env);
  }
  return scheduler.addFiber(element.fiber!).then(() => {
    const fragment = document.createDocumentFragment();
    buildTree(element.vnode!, (fragment as any) as HTMLElement);
    target.appendChild(fragment);
    if (document.body.contains(target)) {
      mountTree(element);
    }
    return element.instance!;
  });
}

function mountTree(element: OwlElement) {
  for (let k in element.children) {
    mountTree(element.children[k]);
  }
  element.isMounted = true;
  if (element.instance.mounted) {
    element.instance.mounted();
  }
}

// -----------------------------------------------------------------------------
// Component creation
// -----------------------------------------------------------------------------

export function createFComponent(fn: FComponent, props: any, env: any): OwlElement {
  let template: string = fn.template;
  const name = fn.name || "Anonymous Function Component";
  if (!template) {
    throw new Error(`Component "${name}" does not have a template defined!`);
  }
  const qtemplate = qweb.getTemplate(template);
  const elem: OwlElement = {
    fiber: null,
    name,
    instance: null,
    isMounted: false,
    vnode: null,
    components: fn.components || {},
    qweb: qtemplate,
    isReady: null,
    children: {},
  };
  core.current = elem;

  if (fn.setup) {
    const result = fn.setup(props, env);
    const prom = result instanceof Promise ? result : Promise.resolve(result);
    elem.isReady = prom.then((result) => {
      let instance = result ? Object.create(result) : {};
      instance.props = props;
      instance.__owl__ = elem;
      instance.env = env;
      elem.instance = instance;
    });
  } else {
    const instance: FInstance = { props, env, __owl__: elem };
    elem.instance = instance;
    elem.isReady = Promise.resolve();
  }

  startComponent(elem);
  return elem;
}

export function createCComponent<C extends typeof Component>(
  C: C,
  props: any,
  env: any
): OwlElement {
  let template: string = C.template;
  if (!template) {
    throw new Error(`Component "${C.name}" does not have a template defined!`);
  }
  const qtemplate = qweb.getTemplate(template);
  const elem: OwlElement = {
    fiber: null,
    name: C.name,
    instance: null,
    isMounted: false,
    vnode: null,
    components: C.components,
    qweb: qtemplate,
    isReady: null,
    children: {},
  };
  core.current = elem;
  const component = new C(props, env) as InstanceType<C>;
  component.__owl__ = elem;
  elem.instance = component;
  elem.isReady = component.willStart();
  startComponent(elem);
  return elem;
}

// -----------------------------------------------------------------------------
// starting Components
// -----------------------------------------------------------------------------

async function startComponent(elem: OwlElement) {
  const vnode = elem.qweb.createRoot();
  vnode.hooks.create = (el) => (elem.instance.el = el);
  const fiber = new Fiber(elem, null, vnode);
  fiber.root.counter++;
  elem.fiber = fiber;
  elem.vnode = vnode;
  await elem.isReady!;
  elem.qweb.render(vnode, elem.instance, elem);
  fiber.root.counter--;
}

export function renderToFiber(elem: OwlElement, parentFiber: Fiber | null): Fiber {
  const newRoot = elem.qweb.createRoot();
  const fiber = new Fiber(elem, parentFiber, newRoot);
  elem.fiber = fiber;
  fiber.root.counter++;
  new Promise((resolve) => {
    elem.qweb.render(newRoot, elem.instance!, elem);
    fiber.root.counter--;
    resolve();
  });
  return fiber;
}
// -----------------------------------------------------------------------------
// render
// -----------------------------------------------------------------------------

export function render(elem: OwlElement): Promise<void> {
  let fiber = renderToFiber(elem, null);
  return scheduler.addFiber(fiber).then(() => {
    patch(elem.vnode!, fiber.vnode!);
    while (fiber.next) {
      fiber = fiber.next;
      patch(fiber.elem.vnode!, fiber.vnode!);
    }
  });
}
