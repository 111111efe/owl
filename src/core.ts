import { BDom, Block, Blocks } from "./bdom";
import { TemplateSet } from "./qweb_compiler";
import { observe } from "./reactivity";

// -----------------------------------------------------------------------------
//  Global templates
// -----------------------------------------------------------------------------

let nextId = 1;
export const globalTemplates = new TemplateSet();

export function xml(strings: TemplateStringsArray, ...args: any[]) {
  const name = `__template__${nextId++}`;
  const value = String.raw(strings, ...args);
  globalTemplates.add(name, value);
  return name;
}

// -----------------------------------------------------------------------------
//  Component
// -----------------------------------------------------------------------------

interface ComponentData {
  bdom: null | BDom;
  render: () => BDom;
  fiber: Fiber | null;
  willStartCB: Function;
  mountedCB: Function;
  isMounted: boolean;
  children: { [key: string]: Component };
  slots?: any;
}

export class Component {
  static template: string;
  __owl__: ComponentData = currentData;
  props: any;
  env: any = currentEnv;

  constructor(props: any) {
    current = this;
    const __owl__ = currentData;
    __owl__.willStartCB = this.willStart.bind(this);
    __owl__.mountedCB = this.mounted.bind(this);
    this.props = props;
  }

  setup() {}
  async willStart(): Promise<void> {}
  async willUpdateProps(props: any): Promise<void> {}
  mounted() {}

  get el(): ChildNode | null {
    const bdom = this.__owl__.bdom;
    return bdom ? bdom.el : null;
  }

  async render(): Promise<void> {
    const fiber = new RootFiber(this.__owl__!);
    scheduler.addFiber(fiber);
    internalRender(this, fiber);
    await fiber.promise;
  }
}

// -----------------------------------------------------------------------------
//  Component Block
// -----------------------------------------------------------------------------

class BComponent extends Block {
  component: Component;
  handlers?: any[];

  constructor(name: string, props: any, key: string, ctx: any) {
    super();
    const parentData: ComponentData = ctx.__owl__;
    let component = parentData.children[key];
    if (component) {
      // update
      const fiber = new ChildFiber(component.__owl__, parentData.fiber!);
      const parentFiber = parentData.fiber!;
      parentFiber.child = fiber; // wrong!
      updateAndRender(component, fiber, props);
    } else {
      // new component
      const components = ctx.constructor.components || ctx.components;
      const C = components[name];
      component = prepare(C, props);
      parentData.children[key] = component;
      const fiber = new ChildFiber(component.__owl__, parentData.fiber!);
      const parentFiber = parentData.fiber!;
      parentFiber.child = fiber; // wrong!
      internalRender(component, fiber);
    }
    this.component = component;
  }

  firstChildNode(): ChildNode | null {
    const bdom = this.component.__owl__.bdom;
    return bdom ? bdom.firstChildNode() : null;
  }

  mountBefore(anchor: ChildNode) {
    this.component.__owl__!.bdom = this.component.__owl__!.fiber!.bdom;
    this.component.__owl__!.bdom!.mountBefore(anchor);
  }
  patch() {
    this.component.__owl__!.bdom!.patch(this.component.__owl__!.fiber!.bdom);
  }
}

export class BComponentH extends BComponent {
  handlers: any[];
  constructor(handlers: number, name: string, props: any, key: string, ctx: any) {
    super(name, props, key, ctx);
    this.handlers = new Array(handlers);
  }

  mountBefore(anchor: ChildNode) {
    super.mountBefore(anchor);
    this.setupHandlers();
  }

  setupHandlers() {
    for (let i = 0; i < this.handlers.length; i++) {
      const handler = this.handlers[i];
      const eventType = handler[0];
      const el = this.component.el!;
      el.addEventListener(eventType, () => {
        const info = this.handlers![i];
        const [, callback, ctx] = info;
        if (ctx.__owl__ && !ctx.__owl__.isMounted) {
          return;
        }
        callback();
      });
    }
  }
}

async function updateAndRender(component: Component, fiber: ChildFiber, props: any) {
  const componentData = component.__owl__;
  componentData.fiber = fiber;
  await component.willUpdateProps(props);
  component.props = props;
  fiber.bdom = componentData.render();
  fiber.root.counter--;
}

Blocks.BComponent = BComponent;
Blocks.BComponentH = BComponentH;

// -----------------------------------------------------------------------------
//  Scheduler
// -----------------------------------------------------------------------------

export class Scheduler {
  tasks: RootFiber[] = [];
  isRunning: boolean = false;
  requestAnimationFrame: Window["requestAnimationFrame"];

  constructor(requestAnimationFrame: Window["requestAnimationFrame"]) {
    this.requestAnimationFrame = requestAnimationFrame;
  }

  start() {
    this.isRunning = true;
    this.scheduleTasks();
  }

  stop() {
    this.isRunning = false;
  }

  addFiber(fiber: RootFiber) {
    this.tasks.push(fiber);
    if (!this.isRunning) {
      this.start();
    }
  }

  /**
   * Process all current tasks. This only applies to the fibers that are ready.
   * Other tasks are left unchanged.
   */
  flush() {
    let tasks = this.tasks;
    this.tasks = [];
    tasks = tasks.filter((fiber) => {
      if (fiber.counter === 0) {
        if (!fiber.error) {
          fiber.complete();
        }
        fiber.resolve();
        return false;
      }
      return true;
    });
    this.tasks = tasks.concat(this.tasks);
    if (this.tasks.length === 0) {
      this.stop();
    }
  }

  scheduleTasks() {
    this.requestAnimationFrame(() => {
      this.flush();
      if (this.isRunning) {
        this.scheduleTasks();
      }
    });
  }
}

const scheduler = new Scheduler(window.requestAnimationFrame.bind(window));

// -----------------------------------------------------------------------------
//  Internal rendering stuff
// -----------------------------------------------------------------------------

class BaseFiber {
  bdom: BDom | null = null;
  error?: Error;
  __owl__: ComponentData;

  child: Fiber | null = null;
  sibling: Fiber | null = null;

  constructor(__owl__: ComponentData) {
    this.__owl__ = __owl__;
  }

  mountComponents() {
    if (this.child) {
      this.child.mountComponents();
    }
    if (this.sibling) {
      this.sibling.mountComponents();
    }
    this.__owl__.mountedCB();
    this.__owl__.isMounted = true;
  }
}

type Fiber = ChildFiber | RootFiber;

class ChildFiber extends BaseFiber {
  bdom: BDom | null = null;
  error?: Error;
  root: RootFiber;
  parent: BaseFiber;

  constructor(__owl__: ComponentData, parent: Fiber) {
    super(__owl__);
    this.parent = parent;
    const root = parent.root;
    root.counter++;
    root.childNumber++;
    this.root = root;
  }
}

class RootFiber extends BaseFiber {
  counter: number = 1;
  childNumber: number = 1;
  root: RootFiber = this;

  resolve!: () => void;
  reject!: (error: Error) => void;
  promise = new Promise((resolve, reject) => {
    this.resolve = resolve;
    this.reject = reject;
  });

  complete() {
    this.__owl__!.bdom!.patch(this.bdom);
  }
}

class MountingFiber extends RootFiber {
  target: HTMLElement | DocumentFragment;

  constructor(__owl__: ComponentData, target: HTMLElement | DocumentFragment) {
    super(__owl__);
    this.target = target;
  }
  complete() {
    const __owl__ = this.__owl__!;
    __owl__.bdom! = __owl__.fiber!.bdom!;
    __owl__.bdom!.mount(this.target);
    if (document.body.contains(this.target)) {
      this.mountComponents();
    }
  }
}

type Env = any;

interface MountParameters {
  env?: Env;
  target: HTMLElement | DocumentFragment;
  props?: any;
}

interface Type<T> extends Function {
  new (...args: any[]): T;
}

let current: Component | null = null;
let currentData: ComponentData;
let currentEnv: any;

export function mount<T extends Type<Component>>(
  C: T | Component,
  params: MountParameters
): Promise<InstanceType<T>>;
export async function mount(C: any, params: MountParameters) {
  if (!(params.target instanceof HTMLElement || params.target instanceof DocumentFragment)) {
    throw new Error("Cannot mount component: the target is not a valid DOM element");
  }
  if (C instanceof Component) {
    // we move component elsewhere
    C.__owl__.bdom!.move(params.target);
    return;
  }
  const { target, props, env } = params;
  currentEnv = env || {};
  const component = prepare(C, props || {});
  const fiber = new MountingFiber(component.__owl__!, target);
  scheduler.addFiber(fiber);
  internalRender(component, fiber);
  await fiber.promise;
  return component;
}

function prepare(C: any, props: any): Component {
  let component: Component;
  let template: string = (C as any).template;
  if (!template) {
    throw new Error(`Could not find template for component "${C.name}"`);
  }
  const __owl__: ComponentData = {
    render: null as any,
    bdom: null,
    fiber: null,
    willStartCB: null as any,
    mountedCB: null as any,
    isMounted: false,
    children: {},
  };
  currentData = __owl__;

  component = new C(props);
  component.setup();
  __owl__.render = globalTemplates.getFunction(template).bind(null, component);
  return component;
}

async function internalRender(c: Component, fiber: Fiber) {
  const __owl__ = c.__owl__!;
  __owl__.fiber = fiber;
  await __owl__.willStartCB();
  fiber.bdom = __owl__.render();
  fiber.root.counter--;
}

// -----------------------------------------------------------------------------
//  hooks
// -----------------------------------------------------------------------------

export function useState<T>(state: T): T {
  const component: Component = current!;
  return observe(state, () => component.render());
}

export function onWillStart(cb: any) {
  const component: Component = current!;
  const prev = currentData!.willStartCB;
  currentData!.willStartCB = () => {
    return Promise.all([prev.call(component), cb.call(component)]);
  };
}

export function onMounted(cb: any) {
  const component: Component = current!;
  const prev = currentData!.mountedCB;
  currentData!.mountedCB = () => {
    prev();
    cb.call(component);
  };
}

// -----------------------------------------------------------------------------
// useRef
// -----------------------------------------------------------------------------

/**
 * The purpose of this hook is to allow components to get a reference to a sub
 * html node or component.
 */
interface Ref<C extends Component = Component> {
  el: HTMLElement | null;
  comp: C | null;
}

export function useRef<C extends Component = Component>(name: string): Ref<C> {
  const __owl__ = currentData!;
  return {
    get el(): HTMLElement | null {
      const val = __owl__.bdom && __owl__.bdom.refs && __owl__.bdom.refs[name];
      return val!;
      // if (val instanceof HTMLElement) {
      //   return val;
      // } else if (val instanceof Component) {
      //   return val.el;
      // }
      // return null;
    },
    get comp(): C | null {
      return null;
      // const val = __owl__.refs && __owl__.refs[name];
      // return val instanceof Component ? (val as C) : null;
    },
  };
}
