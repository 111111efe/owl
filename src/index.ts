import { QWeb } from "./qweb";
import { EventBus } from "./event_bus";
import { Component, PureComponent } from "./component";
import { Store, connect } from "./store";
import * as utils from "./utils";
export const core = {
  QWeb,
  EventBus,
  Component,
  PureComponent,
  Store,
  connect,
  utils
};
