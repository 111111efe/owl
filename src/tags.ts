import { QWeb } from "./qweb/index";

/**
 * Owl Tags
 *
 * We have here a (very) small collection of tag functions:
 *
 * - xml
 *
 * The plan is to add a few other tags such as css, globalcss.
 */

/**
 * XML tag helper for defining templates.  With this, one can simply define
 * an inline template with just the template xml:
 * ```js
 *   class A extends Component {
 *     static template = xml`<div>some template</div>`;
 *   }
 * ```
 */
export function xml(strings, ...args) {
  const name = `__template__${QWeb.nextId++}`;
  const value = String.raw(strings, ...args);
  QWeb.registerTemplate(name, value);
  return name;
}
