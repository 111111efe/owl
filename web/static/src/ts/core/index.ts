export { QWeb } from "./qweb_vdom";
export { EventBus } from "./event_bus";
export { Component } from "./component";
