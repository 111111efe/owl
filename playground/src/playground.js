import { HELLO_WORLD } from "./demos.js";

class Playground extends owl.core.Component {
  inlineTemplate = `
    <div class="playground">
        <div class="left-bar">
            <button t-on-click="runCode">â–¶ Run</button>
            <div class="code-editor" t-ref="editor"></div>
        </div>
        <div class="content">
          <iframe t-ref="iframe"/>
        </div>
    </div>`;

  mounted() {
    this.editor = ace.edit(this.refs.editor);
    this.editor.session.setOption("useWorker", false);
    this.editor.setValue(HELLO_WORLD, -1);
    this.editor.setTheme("ace/theme/monokai");
    this.editor.session.setMode("ace/mode/javascript");
    const doc = this.refs.iframe.contentWindow.document;

    // add owl library
    const owlScript = doc.createElement("script");
    owlScript.src = "libs/owl.js";
    doc.body.appendChild(owlScript);
  }

  runCode() {
    const doc = this.refs.iframe.contentWindow.document;

    // // add owl library
    // const owlScript = doc.createElement("script");
    // owlScript.src = "libs/owl.js";
    // doc.body.appendChild(owlScript);

    const code = this.editor.getValue();
    const script = doc.createElement("script");
    script.type = "text/javascript";
    script.innerHTML = code;
    doc.body.appendChild(script);
    //  var thisDoc = this.contentWindow.document;
    //     if ( ! thisDoc.getElementById(scriptID)) {
    //         var scriptObj = thisDoc.createElement("script");
    //         scriptObj.type = "text/javascript";
    //         scriptObj.id = scriptId;
    //         scriptObj.innerHTML = script;
    //         thisDoc.body.appendChild(scriptObj);
    //     }
    // });
  }
}

document.addEventListener("DOMContentLoaded", async function() {
  const env = {
    qweb: new owl.core.QWeb()
  };

  const playground = new Playground(env);
  playground.mount(document.body);
});
