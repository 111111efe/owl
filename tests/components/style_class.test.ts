import { Component, mount } from "../../src/core";
import { xml } from "../../src/tags";
import { fromName, makeTestFixture, snapshotTemplateCode } from "../helpers";

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

describe("style and class handling", () => {
  test("can set style and class on component", async () => {
    class Test extends Component {
      static template = xml`
          <div style="font-weight:bold;" class="some-class">world</div>
        `;
    }
    await mount(Test, { target: fixture });
    expect(fixture.innerHTML).toBe(`<div style="font-weight:bold;" class="some-class">world</div>`);
    snapshotTemplateCode(fromName(Test.template));
  });
});
