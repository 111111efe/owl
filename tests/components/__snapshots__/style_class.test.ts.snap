// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`style and class handling can set style and class on component 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BText} = Blocks;
  let {elem, toString, withDefault, call, zero, scope, getValues, owner, callSlot} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div style=\\"font-weight:bold;\\" class=\\"some-class\\">world</div>\`);
  }
  
  return ctx => {
    const b1 = new Block1();
    return b1;
  }
}"
`;
